core/src/main/scala/dagr/core/tasksystem/MemoryRetryTask.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package dagr.core.tasksystem
+
+import dagr.core.execsystem.{Resource, TaskExecutionInfo, Memory}
+
+/**
+  * A trait to facilitate retrying a task with more memory.
+  */
+trait MemoryRetryTask extends Task with FixedResources {
+  /** Given the current memory, returns the next memory to retry this task with, or None, if the task
+    * should not be retried.
+    */
+  protected def nextMemory(currentMemory: Memory): Option[Memory]
+
+  /** useSystemMemory true if we should use the system memory as the maximum memory, false to use the heap size. */
+  protected def useSystemMemory: Boolean = true
```

Comment:
I disagree.  An `InJvmTask` could be scheduled with `2G` of memory, try my best to only use `2G` but eventually need more so therefore fail, then require `4G`, and succeed.  Also, we don't physically limit the amount of memory `bwa` uses either, yet we may retry it with parameters to use more memory (I have an old version of `bwa` that can reduce the suffix array compression for faster lookups but more memory).